File: fsharp/queen-attack/QueenAttack.fs
module QueenAttack

let create (position: int * int) = match position with
    | (x, y) when x < 0 || y < 0 -> false
    | (x, y) when x > 7 || y > 7 -> false
    | _ -> true

let canAttack (queen1: int * int) (queen2: int * int) = match (queen1, queen2) with
    | ((x1, _), (x2, _)) when x1 = x2 -> true
    | ((_, y1), (_, y2)) when y1 = y2 -> true
    | ((x1, y1), (x2, y2)) when (abs (x1 - x2)) = (abs (y1 - y2)) -> true
    | _ -> false

File: scala/queen-attack/src/main/scala/QueenAttack.scala
case class Queen(row: Int, col: Int)

object Queen {
  def create(row: Int, col: Int): Option[Queen] = {
    if (isInRange(row) && isInRange(col))
      Some(Queen(row, col))
    else
      None
  }

  private def isInRange(n: Int) = 0 until 8 contains n
}

object QueenAttack {
  private val sameRow = (a: Queen, b: Queen) => a.row == b.row
  private val sameColumn = (a: Queen, b: Queen) => a.col == b.col
  private val sameDiagonal = (a: Queen, b: Queen) => Math.abs(a.row - b.row) == Math.abs(a.col - b.col)

  def canAttack(queen1: Queen, queen2: Queen): Boolean =
    Seq(sameRow, sameColumn, sameDiagonal).exists(_.apply(queen1, queen2))
}